### PR TITLE
bridge: Close channel on capability problems

### PR DESCRIPTION
**Pull request by petervo**
This isn't in use yet but closing the transport was overkill. Closing the channel is enough.

### CODE REVIEW

**Comment by petervo** [comment]
This is not backwards compatible in that if a newer version of cockpit asks an older bridge for a capability it will close the transport, not the channel. However if we rename the parameter ei "caps" or "channel-capabilities" then an older bridge will keep the channel open essentially reporting that it has capabilities that it does not have.

We could work around that by having channels send a message in response to an open command with their capabilities (and options?) and leave it up to the client to decide what to do. But that would also likely cause some backwards compatibility problems.

File: src/bridge/cockpitchannel.c [hunk 1]
```diff
@@ -292,6 +292,37 @@ cockpit_channel_constructed (GObject *object)
...
+  /*
+   * No capabilities are supported yet. When the first one arrives
+   * we'll add code to process it.
+   */
+  gchar **capabilities = NULL;
+  gboolean ret = FALSE;
+
+  if (!cockpit_json_get_strv (options, "capabilities", NULL, &capabilities))
+    {
+      g_message ("got invalid capabilities field in init message");
```
Comment (stefwalter) [inline src/bridge/cockpitchannel.c]: This is actually the "open" message.

**Comment by stefwalter** [comment]
1. It feels like we need to rev the protocol in order to get this right. It's a fundamental feature that is broken.
 2. Or we could just include use "capabilities" and treat this like a bug in those older clients.

Either of the above options have the same effect: The bridge cannot communicate with a newer cockpit.js.

All other solutions involve two way negotiation, which really introduces a whole lot of complexity.

If we are going to do negotiation, then I would prefer to do it globally. To have a way to query a bridge for all the channel payloads it supports and all the various capabilities of those channels. However, doing this by default on every host, would introduce yet more slow down to startup process, which is already in need of streamlining.

**Comment by petervo** [comment]
Fixed the error message.

I agree about avoiding negotiation. I could be wrong.but keeping the argument capabilities feels a little less likely to cause subtle bugs later since it will fail loudly on older bridges. Since we aren't using it now it's hard to predict if that will be better or not, so IDK.

File: src/bridge/cockpitchannel.c [hunk 2]
```diff
@@ -292,6 +292,37 @@ cockpit_channel_constructed (GObject *object)
...
+  if (!cockpit_json_get_strv (options, "capabilities", NULL, &capabilities))
+    {
+      g_message ("got invalid capabilities field in open message");
+      cockpit_channel_close (channel, "protocol-error");
+      goto out;
+    }
+  else if (capabilities && capabilities[0])
+    {
+      g_message ("unsupported capability required: %s", capabilities[0]);
+      cockpit_channel_close (channel, "not-supported");
```
Comment (mvollmer) [inline src/bridge/cockpitchannel.c]: What about delivering the list of supported capabilities in the close message?

**Comment by mvollmer** [comment]
> 1. It feels like we need to rev the protocol in order to get this right. It's a fundamental feature that is broken.
> 2. Or we could just include use "capabilities" and treat this like a bug in those older clients.

Could we define a "channel-caps" capability for this?

Cockpit.js would add this cap to the first channel for a transport, and if the transport is closed because of that, we know that the bridge doesn't support any caps.  If a client asks for caps in that case, cockpit.js rejects that right away without sending a "open" message. 

A bit painful, but that is the price of bug-compatibility.

**Comment by petervo** [comment]
Updated based on our conversation to treat this like a bug. Actually implemented capabilities and adds a tls-certificates capability to httpstream

**Pull request closed without merging**